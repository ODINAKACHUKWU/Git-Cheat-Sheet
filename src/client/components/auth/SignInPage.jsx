import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import Spinner from 'react-md-spinner';
import {
  MDBContainer,
  MDBRow,
  MDBCol,
  MDBCard,
  MDBCardBody,
  MDBModalFooter,
  MDBIcon,
  MDBCardHeader,
  MDBInput,
  MDBBtn
} from 'mdbreact';

import signInAction from '../../actions/creators/userAuthActions';
import { signInValidator } from '../../../../shared/userSharedValidator';
import AppLogo from '../common/AppLogo';

class SignInPage extends Component {
  initialState = {
    data: {
      username: '',
      password: ''
    },
    errors: {}
  };

  state = { ...this.initialState };

  handleOnChange = event => {
    event.persist();
    this.setState(prevState => ({
      ...prevState,
      data: { ...prevState.data, [event.target.name]: event.target.value }
    }));
  };

  handleOnFocus = event => {
    event.persist();
    this.setState(prevState => ({
      errors: { ...prevState.errors, [event.target.name]: '' }
    }));
  };

  handleOnSubmit = event => {
    event.preventDefault();

    const { data } = this.state;
    const { history, signInAction } = this.props;
    const { isValid, errors } = signInValidator(data);

    if (!isValid) {
      this.setState(prevState => ({
        ...prevState,
        errors
      }));
    } else {
      signInAction({ ...data }).then(() => {
        if (!this.props.authError) history.push('/');
      });
    }
  };

  render() {
    const {
      data: { username, password },
      errors
    } = this.state;

    const { isProcessing } = this.props;

    const errorColor = {
      color: '#ff9a9e'
    };

    return (
      <MDBContainer className="mt-4">
        <AppLogo />
        <MDBRow>
          <MDBCol md="6" className="offset-md-3">
            <MDBCard>
              <MDBCardBody>
                <MDBCardHeader className="form-header warm-flame-gradient rounded text-center">
                  <h3 className="my-2">
                    <MDBIcon icon="lock" className="mr-3" />
                    Sign In
                  </h3>
                </MDBCardHeader>
                <form onSubmit={this.handleOnSubmit}>
                  <div className="grey-text">
                    <MDBInput
                      label="username"
                      name="username"
                      value={username}
                      onChange={this.handleOnChange}
                      onFocus={this.handleOnFocus}
                      icon="user-circle-o"
                      group
                      type="text"
                    >
                      {errors.username && (
                        <em
                          className="d-flex justify-content-end"
                          style={errorColor}
                        >
                          <small>{errors.username}</small>
                        </em>
                      )}
                    </MDBInput>
                    <MDBInput
                      label="password"
                      name="password"
                      value={password}
                      onChange={this.handleOnChange}
                      onFocus={this.handleOnFocus}
                      icon="lock"
                      group
                      type="password"
                    >
                      {errors.password && (
                        <em
                          className="d-flex justify-content-end"
                          style={errorColor}
                        >
                          <small>{errors.password}</small>
                        </em>
                      )}
                    </MDBInput>
                  </div>

                  <div className="d-flex justify-content-center mt-4">
                    <MDBBtn
                      color="deep-orange"
                      className="mb-3"
                      type="submit"
                      disabled={isProcessing}
                    >
                      Sign In
                      {isProcessing && (
                        <Spinner
                          className="ml-2"
                          size={18}
                          singleColor="#fff"
                        />
                      )}
                    </MDBBtn>
                  </div>
                </form>
                <MDBModalFooter>
                  <div className="font-weight-light">
                    <p>
                      Not a member?
                      <Link to="/signup" className="ml-2">
                        Sign Up
                      </Link>
                    </p>
                  </div>
                </MDBModalFooter>
              </MDBCardBody>
            </MDBCard>
          </MDBCol>
        </MDBRow>
      </MDBContainer>
    );
  }
}
const mapStateToProps = ({ userAuth: { isProcessing, authError } }) => ({
  isProcessing,
  authError
});

SignInPage.propTypes = {
  authError: PropTypes.string.isRequired,
  isProcessing: PropTypes.bool.isRequired,
  signInAction: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};

export default connect(
  mapStateToProps,
  { signInAction }
)(SignInPage);
