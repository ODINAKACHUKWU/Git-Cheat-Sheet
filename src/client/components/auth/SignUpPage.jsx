import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import Spinner from 'react-md-spinner';
import {
  MDBContainer,
  MDBRow,
  MDBCol,
  MDBCard,
  MDBCardBody,
  MDBModalFooter,
  MDBIcon,
  MDBCardHeader,
  MDBInput,
  MDBBtn
} from 'mdbreact';

import signUpAction from '../../actions/creators/userAuthActions';
import { signUpValidator } from '../../../../shared/userSharedValidator';
import AppLogo from '../common/AppLogo';

class SignUpPage extends Component {
  initialState = {
    data: {
      username: '',
      password: '',
      confirmPassword: ''
    },
    errors: {}
  };

  state = { ...this.initialState };

  handleOnChange = event => {
    event.persist();
    this.setState(prevState => ({
      ...prevState,
      data: { ...prevState.data, [event.target.name]: event.target.value }
    }));
  };

  handleOnFocus = event => {
    event.persist();
    const {
      errors: { password }
    } = this.state;
    if (password && this.checkError(password)) {
      this.setState({ errors: { password: '' } });
    }
    this.setState(prevState => ({
      errors: { ...prevState.errors, [event.target.name]: '' }
    }));
  };

  handleOnSubmit = event => {
    event.preventDefault();

    const { data } = this.state;
    const { history, signUpAction } = this.props;
    const { isValid, errors } = signUpValidator(data);

    if (!isValid) {
      this.setState(prevState => ({
        ...prevState,
        errors
      }));
    } else {
      signUpAction({ ...data }).then(() => {
        if (!this.props.authError) history.push('/');
      });
    }
  };

  checkError = error => {
    if (error.endsWith('match')) return true;
    return false;
  };

  render() {
    const {
      data: { username, password, confirmPassword },
      errors
    } = this.state;

    const { isProcessing } = this.props;

    const errorColor = {
      color: '#ff9a9e'
    };

    return (
      <MDBContainer className="mt-4">
        <AppLogo />
        <MDBRow>
          <MDBCol md="6" className="offset-md-3">
            <MDBCard>
              <MDBCardBody>
                <MDBCardHeader className="form-header warm-flame-gradient rounded text-center">
                  <h3 className="my-2">
                    <MDBIcon icon="unlock-alt" className="mr-3" />
                    Sign Up
                  </h3>
                </MDBCardHeader>
                <form onSubmit={this.handleOnSubmit}>
                  <div className="grey-text">
                    <MDBInput
                      label="username"
                      name="username"
                      value={username}
                      onChange={this.handleOnChange}
                      onFocus={this.handleOnFocus}
                      icon="user-circle-o"
                      group
                      type="text"
                    >
                      {errors.username && (
                        <em
                          className="d-flex justify-content-end"
                          style={errorColor}
                        >
                          <small>{errors.username}</small>
                        </em>
                      )}
                    </MDBInput>
                    <MDBInput
                      label="password"
                      name="password"
                      value={password}
                      onChange={this.handleOnChange}
                      onFocus={this.handleOnFocus}
                      icon="lock"
                      group
                      type="password"
                    >
                      {errors.password && !this.checkError(errors.password) && (
                        <em
                          className="d-flex justify-content-end"
                          style={errorColor}
                        >
                          <small>{errors.password}</small>
                        </em>
                      )}
                    </MDBInput>
                    <MDBInput
                      label="confirm password"
                      name="confirmPassword"
                      value={confirmPassword}
                      onChange={this.handleOnChange}
                      onFocus={this.handleOnFocus}
                      icon="lock"
                      group
                      type="password"
                    >
                      {errors.password && this.checkError(errors.password) && (
                        <em
                          className="d-flex justify-content-end"
                          style={errorColor}
                        >
                          <small>{errors.password}</small>
                        </em>
                      )}
                    </MDBInput>
                  </div>

                  <div className="d-flex justify-content-center mt-4">
                    <MDBBtn
                      color="deep-orange"
                      className="mb-3"
                      type="submit"
                      disabled={isProcessing}
                    >
                      Sign Up
                      {isProcessing && (
                        <Spinner
                          className="ml-2"
                          size={18}
                          singleColor="#fff"
                        />
                      )}
                    </MDBBtn>
                  </div>
                </form>
                <MDBModalFooter>
                  <div className="font-weight-light">
                    <p>
                      Already a member?
                      <Link to="/signin" className="ml-2">
                        Sign In
                      </Link>
                    </p>
                  </div>
                </MDBModalFooter>
              </MDBCardBody>
            </MDBCard>
          </MDBCol>
        </MDBRow>
      </MDBContainer>
    );
  }
}

const mapStateToProps = ({ userAuth: { isProcessing, authError } }) => ({
  isProcessing,
  authError
});

SignUpPage.propTypes = {
  authError: PropTypes.string.isRequired,
  isProcessing: PropTypes.bool.isRequired,
  signUpAction: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};

export default connect(
  mapStateToProps,
  { signUpAction }
)(SignUpPage);
