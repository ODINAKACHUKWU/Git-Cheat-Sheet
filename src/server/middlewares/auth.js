import jwt from 'jsonwebtoken';
import apiResponse from '../helpers/apiResponse';

export default (request, response, next) => {
  const token = request.body.token || request.query.token || request.headers['x-access-token'];

  if (token) {
    // verifies token and checks if expired or invalid
    return jwt.verify(token, process.env.SECRET_KEY, err => {
      if (err)
        return apiResponse.error(
          response,
          401,
          'Authentication failed. Token is invalid or expired'
        );
      return next();
    });
  }
  return apiResponse.error(response, 403, 'Access denied. You are not logged in');
};
