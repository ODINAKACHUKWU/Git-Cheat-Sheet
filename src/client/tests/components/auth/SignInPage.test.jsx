import { SignInPage } from '../../../components/auth/SignInPage';

describe('<SignInPage /> component', () => {
  const setUp = () => {
    const props = {
      authError: '',
      isProcessing: false,
      signInAction: jest.fn(() => Promise.resolve()),
      history: {
        push: jest.fn()
      }
    };
    const state = {
      data: {
        username: '',
        password: ''
      },
      errors: {}
    };
    return {
      shallowWrapper: shallow(<SignInPage {...props} />),
      state,
      props
    };
  };

  const event = {
    target: {
      username: 'new_user',
      pasword: 'password'
    },
    preventDefault: jest.fn(),
    persist: jest.fn()
  };

  it('render without crashing', () => {
    const { shallowWrapper } = setUp();
    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('should display <Loader /> component when `isProcessing` is set to true', () => {
    const { shallowWrapper, props } = setUp();
    shallowWrapper.setProps({
      ...props,
      isProcessing: true
    });
    expect(shallowWrapper.find('MDSpinner')).toHaveLength(1);
  });

  it('invokes handleOnChange method', () => {
    const { shallowWrapper } = setUp();
    const handleOnChangeSpy = jest.spyOn(
      shallowWrapper.instance(),
      'handleOnChange'
    );
    shallowWrapper.instance().handleOnChange(event);
    expect(handleOnChangeSpy).toHaveBeenCalled();
  });

  it('invokes handleOnFocus method', () => {
    const { shallowWrapper } = setUp();
    const handleOnFocusSpy = jest.spyOn(
      shallowWrapper.instance(),
      'handleOnFocus'
    );
    shallowWrapper.instance().handleOnFocus(event);

    expect(handleOnFocusSpy).toHaveBeenCalled();
  });

  it('invokes handleOnSubmit method', () => {
    const { shallowWrapper } = setUp();
    const handleOnSubmitSpy = jest.spyOn(
      shallowWrapper.instance(),
      'handleOnSubmit'
    );
    shallowWrapper.instance().handleOnSubmit(event);

    expect(handleOnSubmitSpy).toHaveBeenCalled();
  });

  it('should signin user when all validations are met', () => {
    const { shallowWrapper, state } = setUp();

    shallowWrapper.setState({
      ...state,
      data: {
        username: 'some_user',
        password: 'password'
      }
    });
    shallowWrapper.find('form').simulate('submit', event);

    expect(shallowWrapper.instance().state.errors).toEqual({});
  });

  it('should not signin user when username or password is fails authentication', () => {
    const { shallowWrapper, state, props } = setUp();

    shallowWrapper.setState({
      ...state,
      data: {
        username: 'some_user',
        password: 'password'
      }
    });

    shallowWrapper.setProps({
      ...props,
      authError: 'Username or passowrd is invalid'
    });
    shallowWrapper.find('form').simulate('submit', event);

    expect(shallowWrapper.instance().state.errors).toEqual({});
  });
});
