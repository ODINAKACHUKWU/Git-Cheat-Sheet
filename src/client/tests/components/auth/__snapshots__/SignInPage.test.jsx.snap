// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignInPage /> component render without crashing 1`] = `
<t
  className="mt-4"
  fluid={false}
  tag="div"
>
  <AppLogo />
  <t
    tag="div"
  >
    <t
      className="offset-md-3"
      lg={null}
      md="6"
      sm={null}
      tag="div"
      xl={null}
      xs={null}
    >
      <t
        tag="div"
      >
        <t
          tag="div"
        >
          <t
            className="form-header warm-flame-gradient rounded text-center"
            tag="div"
          >
            <h3
              className="my-2"
            >
              <t
                className="mr-3"
                icon="lock"
              />
              Sign In
            </h3>
          </t>
          <form
            onSubmit={[Function]}
          >
            <div
              className="grey-text"
            >
              <t
                group={true}
                icon="user-circle-o"
                label="username"
                name="username"
                onChange={[Function]}
                onFocus={[Function]}
                type="text"
                value=""
              />
              <t
                group={true}
                icon="lock"
                label="password"
                name="password"
                onChange={[Function]}
                onFocus={[Function]}
                type="password"
                value=""
              />
            </div>
            <div
              className="d-flex justify-content-center mt-4"
            >
              <t
                className="mb-3"
                color="deep-orange"
                disabled={false}
                tag="button"
                type="submit"
              >
                Sign In
              </t>
            </div>
          </form>
          <t>
            <div
              className="font-weight-light"
            >
              <p>
                Not a member?
                <Link
                  className="ml-2"
                  replace={false}
                  to="/signup"
                >
                  Sign Up
                </Link>
              </p>
            </div>
          </t>
        </t>
      </t>
    </t>
  </t>
</t>
`;
